jurity/recommenders/auc.py

```diff
@@ -0,0 +1,155 @@
+            The DataFrame should contain a minimum of two columns, including self._user_id_column, self._item_id_column,
+            and anything else the metric may need. Each row contains the interaction of one user with one item, and the
+            scores associated with this interaction. There can be multiple interactions per user, and there can be
+            multiple users per DataFrame. However, the interactions for a specific user must be contained within a
+            single DataFrame.
+        predicted_results: pd.DataFrame
+            A pandas DataFrame for the recommended user item interaction data, captured from a recommendation algorithm.
+            The DataFrame should contain a minimum of two columns, including self._user_id_column, self._item_id_column,
+            and anything else the metric may need. Each row contains the interaction of one user with one item, and the
+            scores associated with this interaction. There can be multiple interactions per user, and there can be
+            multiple users per DataFrame. However, the interactions for a specific user must be contained within a
+            single DataFrame.
+        batch_accumulate: bool
+            If specified, this parameter allows you to pass in minibatches of results and accumulate the metric
+            correctly across the batches. This reduces the memory footprint and integrates easily with batched
+            training. If specified, the ``get_score`` function will return a tuple of batch results and accumulated
+            results.
+        return_extended_results: bool
+            Whether the extended results such as the support should also be returned. If specified, the returned results
+            will be of type ``dict``. AUC currently returns ``auc`` and the ``support`` used to calculate AUC.
+
+        Returns
+        -------
+        metric: Union[float, dict, Tuple[float, float], Tuple[dict, dict]]
+            The averaged result(s). The return type is determined by the ``batch_accumulate`` and
+            ``return_extended_results`` parameters. See the examples above.
+        """
+        actual_results = actual_results.set_index([self._user_id_column, self._item_id_column])
+        predicted_results = predicted_results.set_index([self._user_id_column, self._item_id_column])
+        if self.k is not None:
+            sorted_clicks = get_sorted_clicks(predicted_results, self._user_id_column, self.click_column, self.k)
+        else:
+            sorted_clicks = predicted_results
+
+        matches = actual_results.join(sorted_clicks, how='inner', rsuffix='_r')
+        clicks = matches[self.click_column].values
+        predictions = matches[self.click_column + '_r'].values
+        results = np.vstack((clicks, predictions)).T
+
+        return self._accumulate_and_return(results, batch_accumulate, return_extended_results)
```

Comment:
similar to my comment below... `results` is presumably of dimensionality (2 x len(matches) )

in the base class, we do `np.mean(np.concatenate(results))`, which (and I may be wrong here) leads to array of shape 1x 2*len(matches) and then you take average. How is that consistent with the AUC formula where we need to look at the per-observation comparison of predicted & actual? Maybe I am missing something.

Reply:
Note that I overwrite the `_get_results` method in auc.py so `np.mean(np.concatenate(results))` isn't used.

Reply:
Oh absolutely! I missed that one. Thanks!